### PR TITLE
feat: add an AutocompleteWithSuggestions component for reusability

### PR DESCRIPTION
**Pull request by dkoo**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Adds a new component to the Newspack Components library, which is an extension of the `AutocompleteTokenfield` that also renders a list of suggestions to choose from. Takes functions as props for fetching suggestions, fetching saved tokens, and what to do when the `AutocompleteTokenfield` value changes.

This UI pattern is something that @thomasguillot has suggested for several features in progress, including this [Author Profile block](https://github.com/Automattic/newspack-blocks/issues/723) (not yet begun) and the Newspack Listings [parent/child UI](https://github.com/Automattic/newspack-listings/pull/43).

After this lands, we'll want to push a new version of the package to the NPM library.

### How to test the changes in this Pull Request:

It's somewhat hard to test this as an NPM package, but you can test within the Newspack Plugin.

1. In some wizard view, import an `AutocompleteWithSuggestions` component from the `newspack-components` directory. Here's an example of some working code (may need to import additional dependencies such as `apiFetch`, `addQueryArgs`, `decodeEntities`, etc.):

```
<AutocompleteWithSuggestions
	label={ __( 'Search for a post', 'newspack' ) }
	help={ __(
		'Begin typing post title, click autocomplete result to select.',
		'newspack'
	) }
	fetchSavedPosts={ async postIDs => {
		const posts = await apiFetch( {
			path: addQueryArgs( '/wp/v2/posts', {
				per_page: 100,
				include: postIDs.join( ',' ),
				_fields: 'id,title',
			} ),
		} );

		return posts.map( post => ( {
			value: post.id,
			label: decodeEntities( post.title ) || __( '(no title)', 'newspack' ),
		} ) );
	} }
	fetchSuggestions={ async search => {
		const posts = await apiFetch( {
			path: addQueryArgs( '/wp/v2/posts', {
				search,
				per_page: 10,
				_fields: 'id,title',
			} ),
		} );

		// Format suggestions for FormTokenField display.
		const result = posts.reduce( ( acc, post ) => {
			acc.push( {
				value: post.id,
				label: decodeEntities( post.title.rendered ) || __( '(no title)', 'newspack' ),
			} );

			return acc;
		}, [] );
		return result;
	} }
	postType={ 'post' }
	postTypeLabel={ 'Post' }
	maxLength={ 1 }
	onChange={ _value => console.log( _value ) }
	selectedPost={ null }
/>
```

2. In the editor view, observe a component that looks like this, showing your 10 most recent posts in the suggestions list (as returned by the `fetchSuggestions` prop in the above example):

Screen Shot 2021-05-06 at 12 48 10 PM

3. To test as an NPM package, you can fudge it by `cd`ing into the `./assets/components` folder in this repo and running `npm install && run prepublishOnly`. Then, in another repo that already uses components from the `newspack-components` NPM package, copy the contents of the `./assets/components` folder into your `node_modules/newspack-components` folder in the second repo. You can then import like `import { AutocompleteWithSuggestions } from 'newspack-components'` and test using the above example code in an editor page.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by dkoo** [comment]
Actually, @thomasguillot—would you mind if we tackled the [multiselect enhancement](https://github.com/Automattic/newspack-plugin/pull/952#issuecomment-834522156) of this component in a separate PR (and a second, subsequent NPM package release)? I ask because the single-item component is going to be a dependency of [this Campaigns feature in the works](https://github.com/Automattic/newspack-popups/issues/538), and it'd be great to not block that feature in the meantime.

**Comment by thomasguillot** [comment]
> Actually, @thomasguillot—would you mind if we tackled the [multiselect enhancement](https://github.com/Automattic/newspack-plugin/pull/952#issuecomment-834522156) of this component in a separate PR (and a second, subsequent NPM package release)? I ask because the single-item component is going to be a dependency of [this Campaigns feature in the works](https://github.com/Automattic/newspack-popups/issues/538), and it'd be great to not block that feature in the meantime.

Yep, no problem @dkoo

**Review by adekbadek — CHANGES_REQUESTED** [review]
Found some issues: 

1. After the component loads, but before typing anything into the input, selecting a post results in an empty item rendered:

image

2. After selecting a post and typing in characters to select another one, user has to first remove the previously-selected post to insert the post from search results.
3. The UI clearly suggests that multiple items can be selected, while it's not true:

image



> you can test within the Newspack Plugin

We do have a components demo page just for that purpose - I've added it there in 881d816eab353b8c24cbbd29779743a71795cfdb

File: assets/components/src/autocomplete-with-suggestions/index.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import { Button } from '@wordpress/components';
+import { Component } from '@wordpress/element';
+
+/**
+ * External dependencies.
+ */
+import { AutocompleteTokenField } from '../';
+import './style.scss';
+
+class AutocompleteWithSuggestions extends Component {
```
[the remark this answers is not in the document]
Comment (dkoo) [reply assets/components/src/autocomplete-with-suggestions/index.js]: Refactored to a functional component in 74b51b9.

**Comment by dkoo** [comment]
> 1. After the component loads, but before typing anything into the input, selecting a post results in an empty item rendered:

Ah, there was some weirdness in the interaction with the `AutocompleteTokenfield`. In 51bf59f I've extended that component to be able to take an array of tokens that include labels, and if the tokens include labels already, it just displays those instead of trying to fetch the labels.

> 2. After selecting a post and typing in characters to select another one, user has to first remove the previously-selected post to insert the post from search results.

Also fixed in 51bf59f. Now, you should be able to type in a search term, select a suggestion, and it will overwrite the existing value.

> 3. The UI suggests that multiple items can be selected, while it's not true

That's true, but for all use cases of the single-select version of this component, we immediately unmount the component and render something else as soon as a selected value exists, so it hasn't been an issue so far. I'm open to suggestions on how to display a selected state for this component when in single-select mode (maybe we separate the `AutocompleteTokenfield` from the selected tokens as in [Thomas's multi-select design](https://github.com/Automattic/newspack-plugin/pull/952#issuecomment-834522156)?).

File: assets/components/src/autocomplete-with-suggestions/index.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	 * Render a list of suggestions that can be clicked to select instead of searching by title.
+	 */
+	const renderSuggestions = () => {
+		return (
+			<>
+				<p className="newspack-autocomplete-with-suggestions__label">
+					{ __( 'Or, select a recent ', 'newspack' ) + sprintf( ' %s:', postTypeLabel ) }
+				</p>
+				<div className="newspack-autocomplete-with-suggestions__search-suggestions">
+					{ suggestions.map( renderSuggestion.bind( this ) ) }
```
Comment (adekbadek) [inline assets/components/src/autocomplete-with-suggestions/index.js:62]: Binding is not needed anymore:

```suggestion
					{ suggestions.map( renderSuggestion ) }
```

**Review by adekbadek — APPROVED** [review]
👌

**Pull request merged**